tools/src/wyvern/tools/parsing/coreparser/arch/DeclCheckVisitor.java

```diff
@@ -0,0 +1,230 @@
+    // "connectors" maps connector instance name to the connector type
+    private HashMap<String, String> connectors = new HashMap<>();
+    // entrypoints maps entrypoint name to the corresponding component instance name
+    private HashMap<String, String> entrypoints = new HashMap<>();
+    // attachments maps connector instance name to the component ports in the connection
+    private HashMap<String, HashSet<String>> attachments = new HashMap<>();
+    // portdecls maps port name to the port declaration AST node
+    private HashMap<String, ASTPortDecl> portdecls = new HashMap<>();
+
+    public List<ASTComponentDecl> generateDependencyGraph() {
+        // Assemble the dependency graph!
+        HashMap<String, DependencyGraphNode> nodes = new HashMap<>();
+        ArrayList<DependencyGraphNode> noReqNodes = new ArrayList<>();
+        List<ASTComponentDecl> ordered = new LinkedList<>();
+        for (String connector : attachments.keySet()) {
+            HashSet<String> fullports = attachments.get(connector);
+            for (String fullport : fullports) {
+                String component = fullport.substring(0, fullport.indexOf("."));
+                ASTComponentDecl compObj = components.get(component);
+                ASTComponentTypeDecl compType = componentTypes.get(compObj.getType());
+                if (!nodes.containsKey(component)) {
+                    DependencyGraphNode compNode = new DependencyGraphNode(compObj);
+                    nodes.put(component, compNode);
+                }
+                DependencyGraphNode compNode = nodes.get(component);
+                if (compType.getReqs().size() == 0) {
+                    noReqNodes.add(compNode);
+                }
+                for (String reqs : compType.getReqs().values()) {
+                    for (String innerfullport : fullports) {
+                        String innercomponent = innerfullport.substring(0, innerfullport.indexOf("."));
+                        ASTComponentDecl innercompObj = components.get(innercomponent);
+                        ASTComponentTypeDecl innercompType = componentTypes.get(innercompObj.getType());
+                        if (!nodes.containsKey(innercomponent)) {
+                            DependencyGraphNode innercompNode = new DependencyGraphNode(innercompObj);
+                            nodes.put(innercomponent, innercompNode);
+                        }
+                        DependencyGraphNode innercompNode = nodes.get(innercomponent);
+                        for (String provs : innercompType.getProvs().values()) {
+                            if (provs.equals(reqs)) {
+                                compNode.addRequires(innercompNode);
+                                innercompNode.addProvides(compNode);
+                                nodes.put(innercomponent, innercompNode);
+                                nodes.put(component, compNode);
```

Comment:
Why are lines 62 and 63 required? Haven't you already done this in lines 42 and 55?